File: apps/website/src/server/db/seeds/maxima-duracion-de-canciones-en-cd/solution-dynamicProgramming.php
<?php

function maximaDuracionDeCancionesEnCd(array $duraciones, int $capacidad): int {
    $rows = count($duraciones);
    $columns = $capacidad + 1;
    $matrix = array_fill(0, $rows, array_fill(0, $columns, null));
    return aux($matrix, $duraciones, $capacidad, 0);
}

function aux(array &$matrix, array &$duraciones, int $capacidadRestante, int $indice) {
    if ($indice === count($duraciones)) {
        return 0;
    }

    if ($matrix[$indice][$capacidadRestante] === null) {
        if ($duraciones[$indice] > $capacidadRestante) {
            $matrix[$indice][$capacidadRestante] = aux($matrix, $duraciones, $capacidadRestante, $indice + 1);
        } else {
            $matrix[$indice][$capacidadRestante] = max(
                aux($matrix, $duraciones, $capacidadRestante - $duraciones[$indice], $indice + 1) + $duraciones[$indice],
                aux($matrix, $duraciones, $capacidadRestante, $indice + 1)
            );
        }
    }
    return $matrix[$indice][$capacidadRestante];
}

File: apps/website/src/server/db/seeds/maxima-duracion-de-canciones-en-cd/solution-dynamicProgramming.cpp
#include <vector>

using namespace std;

long aux(vector<vector<long>> &matrix, vector<long> &duraciones,
         long capacidadRestante, long indice) {
  if (indice == duraciones.size()) {
    return 0;
  }
  if (matrix[indice][capacidadRestante] == -1) {
    if (duraciones[indice] > capacidadRestante) {
      matrix[indice][capacidadRestante] = aux(matrix, duraciones, capacidadRestante, indice + 1);
    } else {
      matrix[indice][capacidadRestante] = max(
          aux(matrix, duraciones, capacidadRestante - duraciones[indice], indice + 1) + duraciones[indice],
          aux(matrix, duraciones, capacidadRestante, indice + 1)
      );
    }
  }
  return matrix[indice][capacidadRestante];
}

long maximaDuracionDeCancionesEnCd(vector<long> duraciones, long capacidad) {
  long rows = duraciones.size();
  long columns = capacidad + 1;
  vector<vector<long>> matrix(rows, vector<long>(columns, -1));
  return aux(matrix, duraciones, capacidad, 0);
}
